core/src/cgroup.rs

```diff
@@ -1,238 +1,392 @@
-            leftover_time = timeout.saturating_sub(start.elapsed());
+        // We need to check for the existance of cgroup.kill, because
+        // this file does not exist on the root cgroup.
+        let path = self.path.join("cgroup.kill");
+        if !path.exists() {
+            bail!("cannot kill the root cgroup");
         }
+
+        Ok(fs::write(path, "1")?)
     }
 
-    /// Deletes a cgroup by killing it's processes and removing the directory afterwards
-    pub fn delete(&self, timeout: Duration) -> TypedResult<()> {
-        self.kill_all_timeout(timeout)?;
-        for d in WalkDir::new(&self.path)
-            .into_iter()
-            .flatten()
-            .filter(|e| e.file_type().is_dir())
-            .sorted_by(|a, b| a.depth().cmp(&b.depth()).reverse())
-        {
-            std::fs::remove_dir(&d.path()).typ(SystemError::CGroup)?;
-            trace!("Removed {:?}", d.path().as_os_str())
+    /// Returns the path of this cgroup
+    pub fn get_path(&self) -> PathBuf {
+        self.path.clone()
+    }
+
+    /// Returns the path of the event file, which may be polled
+    pub fn get_events_path(&self) -> PathBuf {
+        self.path.join("cgroup.events")
+    }
+
+    /// Kills all processes and removes the current cgroup
+    pub fn rm(&self) -> anyhow::Result<()> {
+        if !is_cgroup(&self.path)? {
+            bail!(
+                "{} is not a valid cgroup",
+                self.path.to_str().unwrap_or("N/A")
+            );
+        }
+
+        // Recursively delete all sub cgroups
+        for entry in fs::read_dir(&self.path)? {
+            let entry = entry?;
+            let meta = entry.metadata()?;
+
+            if meta.is_dir() {
+                // Each sub cgroup is deleted by calling this function recursively
+                let child = Self::import_root(&entry.path())?;
+                child.rm()?;
+            }
         }
+
+        // Remove the actual cgroup
+        self.kill()?;
+        while self.populated()? {}
+        fs::remove_dir(&self.path)?;
+
         Ok(())
     }
 
-    /// Returns the absolute path of this cgroup on the host file system
-    pub fn path(&self) -> PathBuf {
-        self.path.clone()
+    // TODO: Implement functions to fetch the parents and children
+}
+
+/// Returns the first cgroup2 mount point found on the host system
+pub fn mount_point() -> anyhow::Result<PathBuf> {
+    // TODO: This is an awful old function, replace it!
+    procfs::process::Process::myself()?
+        .mountinfo()?
+        .iter()
+        .find(|m| m.fs_type.eq("cgroup2")) // TODO A process can have several cgroup mounts
+        .ok_or_else(|| anyhow!("no cgroup2 mount found"))
+        .map(|m| m.mount_point.clone())
+}
+
+/// Checks if path is a valid cgroup by comparing the device id
+fn is_cgroup(path: &Path) -> anyhow::Result<bool> {
+    let st = statfs::statfs(path)?;
+    Ok(st.filesystem_type() == statfs::CGROUP2_SUPER_MAGIC)
+}
```

Comment:
This could be a static function of the `CGroups` struct, instead of a freestanding static function.

Reply:
Also this function could be public. I see no way for potential misuse.

Reply:
> This could be a static function of the `CGroups` struct, instead of a freestanding static function.

But why? From a semantic point of view I would only use static functions if they are somehow related to that type (e.g. initialization). This is not the case here.